Source/Core/Core/HW/EXI/BBA/BuiltIn.cpp

```diff
@@ -0,0 +1,700 @@
+{
+  SetHardwareInfo(data, dest, src);
+  return std::tuple<Common::EthernetHeader*, Common::IPv4Header*, Common::UDPHeader*>(
+      (Common::EthernetHeader*)data, (Common::IPv4Header*)&data[14],
+      (Common::UDPHeader*)&data[0x22]);
+}
+
+bool CEXIETHERNET::BuiltInBBAInterface::Activate()
+{
+  if (IsActivated())
+    return true;
+
+  active = true;
+  m_in_frame = std::make_unique<u8[]>(9004);
+  m_out_frame = std::make_unique<u8[]>(9004);
+  for (auto& buf : queue_data)
+    buf.reserve(2048);
+  fake_mac = Common::GenerateMacAddress(Common::MACConsumer::BBA);
+  const u32 ip = m_local_ip.empty() ? sf::IpAddress::getLocalAddress().toInteger() :
+                                      sf::IpAddress(m_local_ip).toInteger();
+  m_current_ip = htonl(ip);
+  m_router_ip = (m_current_ip & 0xFFFFFF) | 0x01000000;
+  // clear all ref
+  for (int i = 0; i < std::size(network_ref); i++)
+  {
+    network_ref[i].ip = 0;
+  }
+
+  return RecvInit();
+}
+
+void CEXIETHERNET::BuiltInBBAInterface::Deactivate()
+{
+  // Is the BBA Active? If not skip shutdown
+  if (!IsActivated())
+    return;
+  // Signal read thread to exit.
+  m_read_enabled.Clear();
+  m_read_thread_shutdown.Set();
+  active = false;
+
+  // kill all active socket
+  for (int i = 0; i < std::size(network_ref); i++)
+  {
+    if (network_ref[i].ip != 0)
+    {
+      network_ref[i].type == IPPROTO_TCP ? network_ref[i].tcp_socket.disconnect() :
+                                           network_ref[i].udp_socket.unbind();
+    }
+    network_ref[i].ip = 0;
+  }
+
+  // Wait for read thread to exit.
+  if (m_read_thread.joinable())
+    m_read_thread.join();
+}
+
+bool CEXIETHERNET::BuiltInBBAInterface::IsActivated()
+{
+  return active;
+}
+
+void CEXIETHERNET::BuiltInBBAInterface::WriteToQueue(const u8* data, int length)
+{
+  queue_data[queue_write].resize(length);
+  std::memcpy(&queue_data[queue_write][0], data, length);
+  if (((queue_write + 1) & 15) == queue_read)
+  {
+    return;
+  }
+  queue_write = (queue_write + 1) & 15;
+}
+
+void CEXIETHERNET::BuiltInBBAInterface::HandleARP(Common::EthernetHeader* hwdata,
+                                                  Common::ARPHeader* arpdata)
+{
+  std::memset(m_in_frame.get(), 0, 0x30);
+  Common::EthernetHeader* hwpart = (Common::EthernetHeader*)m_in_frame.get();
+  *hwpart = Common::EthernetHeader(ARP_PROTOCOL);
+  hwpart->destination = *(Common::MACAddress*)&m_eth_ref->mBbaMem[BBA_NAFR_PAR0];
+  hwpart->source = fake_mac;
+
+  Common::ARPHeader* arppart = (Common::ARPHeader*)&m_in_frame[14];
+  Common::MACAddress bba_mac = *(Common::MACAddress*)&m_eth_ref->mBbaMem[BBA_NAFR_PAR0];
+  if (arpdata->target_ip == m_current_ip)
+  {
+    // game asked for himself, reply with his mac address
+    *arppart = Common::ARPHeader(arpdata->target_ip, bba_mac, m_current_ip, bba_mac);
+  }
+  else
+  {
+    *arppart = Common::ARPHeader(arpdata->target_ip, fake_mac, m_current_ip, bba_mac);
+  }
+
+  WriteToQueue(&m_in_frame[0], 0x2a);
+}
```

Comment:
Like, take this function for example. Regardless of if it works or not, there's several issues here that make it difficult to understand and possibly change in the future:

- Magic numbers. Why is this memsetting 0x30 bytes to zero, but then only writing 0x2a of them? What do those bytes represent, even?
- Lots of type punning that violates the C++ spec.
- Even if it were allowed, it's hard to mentally process. The writes to `hwpart` and `arppart` do indirectly affect the data stored in `m_in_frame`!
- And last but not least, why is this even writing to `m_in_frame`? Does something else rely on those 0x30 bytes being written? If so, where, and why is there no comment explaining that? If not, why is this not just a 0x30 bytes array on the stack instead?

This code is (unfortunately) full of things like this. These need to be fixed before it can be merged, or we're just creating headaches for future maintainers of this project.

Reply:
i understands all those issue, sadly im far from being a guru of C++ and the code im digging in right now is far from your todays standard. Since it look like there is still tones of changes to do, I will push one last commit for most of what was requested today, then it would be nice if someone more knoledgeable in C++ could take over.... 
I can provide information on what it does, like in this case 0x30 fill is to clear the memory to ensure you get the proper value, the original reply was 0x30 and got reduced in the revisions, so 0x30 can now beacome 0x2a, i will review the size.

Most of the code was based on the tapdevice code as a starting point, why i use m_in_frame? beacause it was like that in the tap or xlink device.  Im not saying this is the the right way, and i could of just reallocate a new buffer every time, just tought it was faster cpu side to not free/alloc memory, and it looked like this was used like that elsewhere too.

Reply:
Yeah, that's totally fair! The TAP code in particular is probably 10 years old and our standards were very different back then. And genuinely, I'm impressed you managed to get this to work at all if you're not very C++ knowledgeable.

If you want, someone else can definitely take this over and fix it up for you as long as you're willing to answer any questions we might have about the code.

Reply:
All valid points guys, thank you all for your hard work.

Reply:
I can provide my Discord id to the team or the one taking over. I will be happy to awnser and fallow the result to learn more about c++.